XRDDEV-1419 Kubernetes Documentation (#46)
nordic-institute/X-Road-Security-Server-sidecar: doc/kubernetes_security_server_sidecar_user_guide.md

 Date       | Version | Description                                                     | Author
 ---------- | ------- | --------------------------------------------------------------- | --------------------
 05.01.2021 | 1.0     | Initial version                                                 | Alberto Fernandez Lorenzo

# Table of Contents
* [1 Introduction](#1-introduction)
   * [1.1 Target Audience](#11-target-audience)
* [2 Deployment Options](#2-deployment-options)
   * [2.1 Single Pod Deployment with internal database](#21-single-pod-deployment-with-internal-database)
   * [2.2 Single Pod Deployment with external database](#22-single-pod-deployment-with-external-database)
   * [2.3 Multiple Pods using a Load Balancer](#23-multiple-pods-using-a-load-balancer)
* [3 X-Road Security Server Sidecar images for Kubernetes](#3-x-road-security-server-sidecar-images-for-kubernetes)
* [4 Installation](#4-installation)
   * [4.1 Minimum system requirements](#41-minimum-system-requirements)
   * [4.2 Prerequisites to Installation](#42-prerequisites-to-installation)
   * [4.3 Network configuration](#43-network-configuration)
   * [4.4 Reference Data](#44-reference-data)

>    * [4.2 Installation Instructions](#42-installation-instructions)

Reviewer: Chapter numbering seems to be bit off... after 4.4 should come 4.5

  volumeMode: Filesystem
  accessModes:
  - <volume access mode>
  persistentVolumeReclaimPolicy: Retain
  csi:
    driver: efs.csi.aws.com
    volumeHandle: <efs volume id>
```

- It is possible to verify the creation and get the volume info by running (Authentication through the AWS CLI it's required) (**reference Data: 3.11**):
```
aws efs describe-mount-targets --file-system-id <efs volume id>
```

- It is possible to mount the volume in any Cluster Node instance by running from the Cluster Node Instance where the volume is going to be mounted (**reference Data: 3.11, 3.22**):
```
sudo mount -t nfs -o nfsvers=4.1,rsize=1048576,wsize=1048576,hard,timeo=600,retrans=2,noresvport <efs volume id>.efs.<cluster region>.amazonaws.com:/ <local folder>
```
##### 4.2.3.2.4 AWS EBS vs AWS EFS
The table below shows us a comparison with the main differences and when it is recommended to use each one to choose the best alternative between AWS EBS and EFS volumes.

|     | **AWS EBS**                                          | AWS EFS                                   |
| --- | ---------------------------------------------------- | ----------------------------------------- |

> | Price | General Purpose SSD(gp2) Volumes:  $0.10 per GB-month of provisioned storage.EBS Snapchots: $0.05 per GB-month of data stored. | Standard Storage (GB-Month) : $0.30. |

Reviewer 2: This is still pending

| Durability | 20 times more reliable than normal hard disks | Highly durable (No public SLA) |
| Availability | 99.99% available.Cannot withstand availability zone failure without snapshots. | Highly available service.  (No public SLA)Every file system object is redundantly stored across multiple Availability Zones so it can survive one AZ failure. |
| Backup | Provides point in time snapshots, witch are backed up to AWS S3 and can be copied across the regions. | EFS doesn’t support any backup mechanism we need to setup backup manually. We must use AWS backup. |
| When to use | Single node scenario.Reduce costs.We want a faster option.We don't need to worry about scalability.The amount of data is not that big.Create and recover backups in an easy way | Multiple node scenario.We need concurrent access.We want automatically scalability.The amount of data is big. |

#### 4.2.3.3 Manage Volumes
Once the volume manifest is created, we can deploy it by running:
```
kubectl apply -f /path/to/pv_file.yaml
```
List the PV by running (**reference Data: 3.1**):
```
kubectl get pv -n <namespace name>
```

> List the PVC by running (The PVC status should be boud, it means that the claim is attached to a PV):

Reviewer 2: ```suggestion
List the PVC by running (The PVC status should be bound, which means that the PVC is attached to a PV):
```

```
kubectl get pvc
```
Delete the PV or PVC by running:
```

> kubectl delete -f /path/to/file.yaml.

Reviewer 2: ```suggestion
kubectl delete -f /path/to/file_name.yaml.
```

```

#### 4.2.3.4 Mount the Volume to a Pod

> To mount the volume on a Pod, it is required to have a PVC bounded to a PV, then modify the Pod manifest, in this case the manifest defined in the step [2.1 Single Pod Deployment with internal database](#21-single-pod-deployment-with-internal-database) will be modified by mapping the volume to the xroad configuration `etc/xroad` (**reference Data: 3.4, 3.12, 3.13**):

Reviewer 2: ```suggestion
To mount the volume to a Pod, we should make sure to have a PVC bounded to a PV and modify the Pod manifest, for example, the manifest defined in the step [2.1 Single Pod Deployment with internal database](#21-single-pod-deployment-with-internal-database), by mapping the volume to the xroad configuration `etc/xroad` as follows (**reference data: 3.4, 3.12, 3.13**):
```

``` yaml
[...]
spec:
  volumes:
  - name: <manifest volume name>
    persistentVolumeClaim:
      claimName: <pvc name>
   containers:
   - name: <container name>
     image: niis/xroad-security-server-sidecar:<image tag>
     imagePullPolicy: "Always"
     volumeMounts:
     - name: <manifest volume name>
       mountPath: /etc/xroad/
[...]
```

#### 4.2.3.5 Persistent Volume permissions

> When you mount a volume to a pod, by default it always gets mounted with the permission of root:root.

Reviewer 2: ```suggestion
When mounting a volume to a Pod, by default it always gets mounted with the permission of root:root.
```

> This could be a problem when mapping the volume to local folders like the configuration folder `etc/xroad` or the message-log backup `var/lib/xroad` that requires xroad:xroad (UID: 999, GID:999) permission.

Reviewer 2: ```suggestion
This could be a problem when the volume is mapped to local paths such as `etc/xroad` or `var/lib/xroad`, which both require xroad:xroad (UID: 999, GID:999) permission.
```

> Kubernetes allow us to deploy the container as another user using Security Context, but this doesn't work because Security Server Sidecar requires to be run by the root user.

Reviewer 2: ```suggestion
Kubernetes allows deploying the container as another non-root user using Security Context configuration. However, this approach is not yet possible since the Security Server Sidecar requires to be run by the root user.
```

> It is possible to change manually the permission setting the UID and GID in the folder of the Cluster node instance where the volume is mounted.

Reviewer 2: ```suggestion
A possible solution is to change manually the permissions by setting the UID and GID in the folder of the Cluster node instance where the volume is mounted.
```

> Another possible is use Kubernetes Init Containers for setting the permissions to the volume folder, wich launches before the main container. An example for change the permissions to the `etc/xroad` volume folder could be (**reference Data: 3.4, 3.13**):

Reviewer 2: ```suggestion
Another possible solution is to use Kubernetes Init Containers for setting the correct permissions to the volume folder which launches before the main container. Below is an example configuration for changing the permissions to the `etc/xroad` volume folder (**reference data: 3.4, 3.13**):
```

``` yaml
[...]
spec:
  volumes:
  - name: <manifest volume name>
    persistentVolumeClaim:
      claimName: <pvc name>
  initContainers:
  - name: volume-hack
    image: busybox:latest
    command: ["sh","-c","chown  999:999 /etc/xroad"]
    volumeMounts:
    - name: <manifest volume name>
      mountPath: /etc/xroad

[...]
```

> Note (1) References: https://stackoverflow.com/questions/43544370/kubernetes-how-to-set-volumemount-user-group-and-file-permissions

Reviewer 2: ```suggestion
Note (1) References: [StackOverflow](https://stackoverflow.com/questions/43544370/kubernetes-how-to-set-volumemount-user-group-and-file-permissions)
```

>                      https://serverfault.com/questions/906083/how-to-mount-volume-with-specific-uid-in-kubernetes-pod

Reviewer 2: ```suggestion
                     [Serverfault](https://serverfault.com/questions/906083/how-to-mount-volume-with-specific-uid-in-kubernetes-pod)
```

### 4.2.4 Kubernetes Secrets
### 4.2.4.1 Store keys in Secrets
Kubernetes Secrets allows us to store and manage sensitive information.

For the [2.3 Multiple Pods using a Load Balancer](#23-multiple-pods-using-a-load-balancer) scenario you need to create a Kubernetes secret, this secret will store the ssh keys used by the Secondary Pods to synchronize the configuration with the Primary Pod.
If you don't have an ssh key you can create one by running:
```
ssh-keygen -f /path/to/.ssh/
```
Then create a Kubernetes Secret for storing the ssh keys by running (**reference Data: 3.1, 3.14**);
```
kubectl create secret generic <secret name> --from-file=private-key=/path/to/.ssh/id_rsa --from-file=public-key=/path/to/.ssh/id_rsa.pub --namespace=<namespace name>
```

#### 4.2.4.2 Secrets for environmental variables
This example shows how to create a secret for the Security Server Sidecar environment variables with sensitive data.

> - Create a manifest file called for example 'secret-env-variables.yaml' and fill it with the desired values of the environment variables ( **reference Data: 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 1.10**):

Reviewer 2: ```suggestion
- Create a manifest file called for example 'secret-env-variables.yaml' and fill it with the desired values of the environment variables ( **reference Data: 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 1.10, 3.1**):
```

```bash
apiVersion: v1
kind: Secret
metadata:
  name: secret-sidecar-variables
  namespace: <namespace_name>
type: Opaque
stringData:
  XROAD_TOKEN_PIN: "<token pin>"
  XROAD_ADMIN_USER: "<admin user>"
  XROAD_ADMIN_PASSWORD: "<admin password>"
  XROAD_DB_HOST: "<database host>"
  XROAD_DB_PWD: "<database password>"
  XROAD_DB_PORT: "database port"
  XROAD_LOG_LEVEL: "<xroad log level>"
```
- Apply the manifest:
```bash
$ kubectl apply -f secret-env-variables.yaml
```

#### 4.2.4.3 Consume secrets

> The Secrets that store keys can be consumed in a similar way to volumes, for this you will have to include the Secret in the definition of volumes within the Pod deployment manifest, select the key and assign it permissions, then mount the volume in a folder on the container (**reference Data: 3.13, 3.14**):

Reviewer 2: ```suggestion
The Secrets that store keys can be consumed in a similar way to volumes. To do this, you will have to include the Secret in the definition of volumes within the Pod deployment manifest, select the key and assign permissions to it, then mount the volume in a folder on the container (**reference Data: 3.13, 3.14**):
```

containers:
livenessProbe:
  httpGet:
   path: /
   port: 80
  initialDelaySeconds: 100
  periodSeconds: 10
  successThreshold: 1
  failureThreshold: 5
  [...]
  ```

#### 4.2.5.3 Startup probes
The startup probes indicate whether the application within the container is started. All other probes are disabled if a startup probe is provided until it succeeds.

Startup probes are useful for Pods that have containers that take a long time to come into service. This is not really useful in the Sidecar pod because it takes to short to start.
In a different scenario where the Sidecar container would take a long time to start, the startup probe can be used in combination with the liveness probe, so that it waits until the startup probe has succeeded before starting the liveness probe. The tricky part is to set up a startup probe with the same command, HTTP or TCP check, with a failureThreshold * periodSeconds long enough to cover the worse case startup time.

 ```bash
 [...]
containers:
livenessProbe:
 httpGet:
  path: /
  port: 80
 periodSeconds: 10
 successThreshold: 1
 failureThreshold: 50
 [...]
 ```

### 4.2.6 Multiple Pods using a Load Balancer deployment
#### 4.2.6.1 Prerequisites

> - A Persitent Volume Claim bounded to a Persitent Volume for store the Primary Pod configuration [4.2.3 Kubernetes Volumes](#423-Kubernetes-volumes).

Reviewer 2: ```suggestion
- A Persistent Volume Claim is bounded to a Persistent Volume to store the Primary Pod configuration [4.2.3 Kubernetes Volumes](#423-Kubernetes-volumes).
```

> - A Kubernetes Secret with a ssh key pair stored [4.2.4 Kubernetes Secrets](#424-Kubernetes-secrets).

Reviewer 2: ```suggestion
- A Kubernetes Secret with an ssh key pair is stored [4.2.4 Kubernetes Secrets](#424-Kubernetes-secrets).
```

     imagePullPolicy: "Always"
     volumeMounts:
     - name: <manifest volume name>
       mountPath: /etc/xroad/
     - name: <manifest volume name_2>
       mountPath: "/etc/.ssh/"
     env:
     - name: XROAD_TOKEN_PIN
       value: "<token pin>"
     - name: XROAD_ADMIN_USER
       value: "<admin user>"
     - name: XROAD_ADMIN_PASSWORD
       value: "<admin password>"
     - name: XROAD_LOG_LEVEL
       value: "<xroad log level>"
     - name: XROAD_DB_HOST
       value: "<database host>"
     - name: XROAD_DB_PORT
       value: "<database port>"
     - name: XROAD_DB_PWD
       value: "<xroad db password>"
     - name: XROAD_DATABASE_NAME
       value: "<database name>"
     ports:
     - containerPort: 4000
     - containerPort: 5500
     - containerPort: 5577
     - containerPort: 5588
     - containerPort: 22
```

The manifest has two Kubernetes objects:
- A Headless Service, this service is used so that the Secondary Pods can connect to the primary one via ssh through a fixed DNS, this DNS will be "&lt;service nam&gt; .&lt;namespace name&gt; .svc.cluster.local"  because the private IP of the primary Pod can change each time it is recreated.

> This service has no open port since is not required any communication from outside.

Reviewer 2: ```suggestion
This service has no open ports since it does not require any communication to or from the outside.
```

- A Pod with the primary image of the Security Server Sidecar, as image tag we can choose between the "primary" or "primary-slim" described in [3 X-Road Security Server Sidecar images for Kubernetes](#3-x-road-security-server-sidecar-images-for-Kubernetes).

> The Pod defines two volumes, one for store the secret public key described in [4.2.4 Kubernetes Secrets](#424-Kubernetes-secrets) and a volume to store the `etc/xroad` configuration. It is possible to choose between the volumes described in [4.2.3 Kubernetes Volumes](#423-Kubernetes-volumes).

Reviewer 2: ```suggestion
The Pod defines two volumes: one volume to store the secret public key described in [4.2.4 Kubernetes Secrets](#424-Kubernetes-secrets), and a second volume to store the `etc/xroad` configuration. It is possible to choose between the volumes described in [4.2.3 Kubernetes Volumes](#423-Kubernetes-volumes).
```

Once the Primary Pod is deployed we need to configure it (register in the Central Server, create the certificates...) following the [User Guide](https://github.com/nordic-institute/X-Road-Security-Server-sidecar/blob/master/doc/security_server_sidecar_user_guide.md#43-configuration).

> Once the configuration is ready, it is possible to verify by running a Healthcheck from inside the internal network, checking that the result is OK:

Reviewer 2: ```suggestion
Once the configuration is ready, it is possible to verify the installation by running a Healthcheck to the Pod running the Security Server Sidecar container from the internal network and checking that the result is OK:
```

```
curl -i <private pod ip>:5588
```

#### 4.2.6.3 Secondary Pods installation

> An example of how to install the Secondary Pod is shown in the manifest below (**reference Data: 3.1, 3.3, 3.4, 3.12, 3.13, 3.14,3.15, 3.17, 3.18, 3.19, 3.20, 1.4, 1.5, 1.6, 1.10**):

Reviewer 2: ```suggestion
An example of how to install the Secondary Pod is shown in the manifest below (**reference Data: 3.1, 3.2, 3.4, 3.12, 3.13, 3.14, 3.15, 3.17, 3.18, 3.19, 3.20, 1.4, 1.5, 1.6, 1.10**):
```

          secretName: <secret name>
          items:
          - key: private-key
            path: id_rsa
            mode: 0644
          - key: public-key
            path: id_rsa.pub
            mode: 0644
      containers:
      - name: <container name>
        image: niis/xroad-security-server-sidecar:<image tag>
        imagePullPolicy: "Always"
        volumeMounts:
        - name: <manifest volume name>
          mountPath: "/root/.ssh/"
        env:
        - name: XROAD_TOKEN_PIN
          value: "<token pin>"
        - name: XROAD_ADMIN_USER
          value: "<admin user>"
        - name: XROAD_ADMIN_PASSWORD
          value: "<admin password>"
        - name: XROAD_LOG_LEVEL
          value: "<xroad log level>"
        - name: XROAD_PRIMARY_DNS
          value: "<primary DNS>"
        readinessProbe:
          httpGet:
            path: /
            port: 5588
          initialDelaySeconds: 200
          periodSeconds: 10
          successThreshold: 1
          failureThreshold: 1
        ports:
        - containerPort: 5500
        - containerPort: 5577
        - containerPort: 5588
        - containerPort: 22
```

The manifest has two Kubernetes objects:

> - An NLB (Network Load Balancer) which will be in charge of redirecting the traffic between the different Secondary pods. The ClusterIP property is optional and is used to assign a fixed private IP to the Load Balancer, this can help you in the configuration of the Security Server Sidecar at the central server, if not assigned, a new private IP will be assigned to the Load Balancer in each deployment.

Reviewer 2: ```suggestion
- An NLB (Network Load Balancer) which will be in charge of redirecting the traffic between the different Secondary pods. The ClusterIP property is optional and is used to assign a fixed private IP to the Load Balancer, this can help you in the configuration of the Security Server Sidecar at the central server. If the ClusterIP is not assigned, a new private IP will be assigned to the Load Balancer in each deployment.
```

> It has the required ports "5500" "5577" for the messages communication between Security Servers open.

Reviewer 2: ```suggestion
It has the required ports "5500" and "5577" to exchange messages between Security Servers open.
```

> - A DeploymentControll for deploy the "n" numbers of Pods with the image for the Secondary Pods, as image tag we can choose between the "secondary" or "secondary-slim" described in [3 X-Road Security Server Sidecar images for Kubernetes](#3-x-road-security-server-sidecar-images-for-Kubernetes).

Reviewer 2: ```suggestion
- A DeploymentControl for deploying the "n" numbers of Pods with the image for the Secondary Pods. As image tag, we can choose between the "secondary" or "secondary-slim" described in [3 X-Road Security Server Sidecar images for Kubernetes](#3-x-road-security-server-sidecar-images-for-Kubernetes).
```

> The Pods have one volume for store the public and private ssh keys defined on [4.2.4 Kubernetes Secrets](#424-Kubernetes-secrets) require for the synchronization with the Primary Pod via ssh.

Reviewer 2: ```suggestion
The Pods have one volume to store the public and private ssh keys defined on [4.2.4 Kubernetes Secrets](#424-Kubernetes-secrets) which are required for the synchronization with the Primary Pod via ssh.
```

> It is not required to add a new volume for store the configuration since all the secondary Pods synchronize the configuration with the Primary, but could be use to make sure that all the Secondary Pods have the same configuration in case any fail to synchronize.

Reviewer 2: ```suggestion
It is not required to add a new volume to store the Secondary Pod configuration, since all the secondary Pods synchronize the configuration with the Primary, but it could be useful to make sure that all the Secondary Pods have the same configuration in case any of them fail to synchronize.
```

The Secondary Pods also have a "ReadinessProbe", this test will run a healthcheck every 10 seconds, starting 200 seconds after deployment, as long as the healthcheck result is not positive, the Pod status will remain in "NotReady" and will not be included in the redirection of the Load Balancer.

> After the manifest is deployed we can scale the Secondary Pods by running:

Reviewer 2: Reference data missing

```
kubectl scale -n <namespace name> --replicas=<number replicas> deployment/<pod name>
```

The Secondary Pods will synchronize the configuration at initialization and through a cron job that runs every minute. Once the configuration is sync, the secondary Pods can process the messages independently of the primary one, this means that if the primary Pods crashes, the cron that synchronizes the configuration will fail but the Secondary Pods can continue to process the messages.

# 5 Backup and Restore

> The backup system of the Security Servers described [here](https://github.com/nordic-institute/X-Road/blob/develop/doc/Manuals/ug-ss_x-road_6_security_server_user_guide.md#13-back-up-and-restore)  is still valid for the installation using Kubernetes.

Reviewer 2: ```suggestion
The backup system of the Security Servers described in the [User Guide](https://github.com/nordic-institute/X-Road/blob/develop/doc/Manuals/ug-ss_x-road_6_security_server_user_guide.md#13-back-up-and-restore) is also valid for the installation using Kubernetes.
```

If your Kubernetes deployment uses volumes to store the configuration, you can back up each volume.
As described in the [4.2.3 Kubernetes Volumes](#423-Kubernetes-volumes) section we will have 3 types of volume, each one with a way of create a backup:

> - AWS Elastic Block Store: Backups can be stored by creating a snapshot of the volume from the AWS admin UI, these snapshots can be configured automatically also it is possible to restore the snapshots into the volume.

Reviewer 2: ```suggestion
- AWS Elastic Block Store: Backups can be stored by creating a snapshot of the volume from the AWS admin UI. These snapshots can be configured automatically and it is possible to restore the snapshots into the volume.
```

- AWS Elastic File System: It is possible to create and restore backups of this volume using [AWS Backup](https://docs.aws.amazon.com/efs/latest/ug/awsbackup.html).
- Host path: This volume does not have any tool to make backup copies, if necessary it will be required to do them manually by copying the contents of the volume.

# 6 Monitoring
**Amazon CloudWatch** monitors your Amazon Web Services (AWS) resources and the applications you run on AWS in real time. You can use CloudWatch to collect and track metrics, which are variables you can measure for your resources and applications.
The CloudWatch home page automatically displays metrics about every AWS service you use.
You can create alarms that watch metrics and send notifications or automatically make changes to the resources you are monitoring when a threshold is breached.

> For more information about Cloudwatch check [here](https://docs.aws.amazon.com/cloudwatch/index.html).

Reviewer 2: ```suggestion
For more information about CloudWatch check the [Amazon CloudWatch documentation](https://docs.aws.amazon.com/cloudwatch/index.html).
```

> **CloudWatch container insights** is a tool available for AWS EKS that we can use to  collect, aggregate, and summarize metrics and logs from your containerized applications and microservices.

Reviewer 2: ```suggestion
**CloudWatch container insights** is a tool available for AWS EKS that we can use to collect, aggregate, and summarize metrics and logs from your containerized applications and microservices.
```

```
curl https://raw.githubusercontent.com/aws-samples/amazon-cloudwatch-container-insights/latest/k8s-deployment-manifest-manifests/deployment-mode/daemonset/container-insights-monitoring/quickstart/cwagent-fluentd-quickstart.yaml | sed "s/{{cluster_name}}/<cluster name>/;s/{{region_name}}/<cluster region>/" | kubectl apply -f -
```
- Verify the installation by running:
```
kubectl get pods -n amazon-cloudwatch
```
- View the logs of the cloudwatch agent by running (**reference Data: 3.23**):
```
kubectl logs <cloudwatch agent name> -n amazon-cloudwatch
```
- It is possible to delete the deployment by running:
```
curl https://raw.githubusercontent.com/aws-samples/amazon-cloudwatch-container-insights/latest/k8s-deployment-manifest-manifests/deployment-mode/daemonset/container-insights-monitoring/quickstart/cwagent-fluentd-quickstart.yaml | sed "s/{{cluster_name}}/<cluster-name>/;s/{{region_name}}/<cluster-region>/" | kubectl delete -f -
```

Once the agent is installed, you can see the cluster metrics, create alerts ... from the AWS admin console on the CloudWatch => Container Insights screen.

# 7 Version update

> It will be possible to update the version of the Security Server Sidecar by re-displaying the image with a higher version in the image tag.

Reviewer 2: ```suggestion
It is possible to update the version of the Security Server Sidecar by re-deploying the image with a higher version in the image tag.
```

> In case of the scenario [2.3 Multiple Pods using a Load Balancer](#23-multiple-pods-using-a-load-balancer) and it is required to no disrupt the connection while the version is updating, it is required to take several steps:

Reviewer 2: ```suggestion
In case of the scenario [2.3 Multiple Pods using a Load Balancer](#23-multiple-pods-using-a-load-balancer), if we do not want to disrupt the connection during the version update, it is required to take several steps:
```

> - First, it is required to stop in the Secondary Pods, the cron job service that handles synchronization with the Primary Pod, it is possible to do that by running in a console for each Secondary Pod (**reference Data: 3.2**):

Reviewer 2: ```suggestion
- First, we should stop the cron job service running in the Secondary Pods that handles synchronization with the Primary Pod by running in a console for each Secondary Pod (**reference Data: 3.2**):
```

```
kubectl exec <pod name> -n $1 supervisorctl stop cron
```
Note (1) It is possible to use some script like this to stop the service for all the Pods in a namespace (**reference Data: 3.1**):
```
for pod in $(kubectl get po -n <namespace name> -oname | awk -F "/" '{print $2}');
   do kubectl exec $pod -n $1 supervisorctl stop cron;
done;
```

> - Update the Primary Pod by open a shell for the Primary Pod container by running:

Reviewer 2: ```suggestion
- Update the Primary Pod by opening a Shell to the Primary Pod container by running:
```

```
kubectl exec -it <pod name> -n <namespace name> bash
```
Stop the Services:
```
supervisorctl stop all
```
Add the new version repository key:
```
echo "deb https://artifactory.niis.org/xroad-release-deb bionic-current main" >/etc/apt/sources.list.d/xroad.list && apt-key add '/tmp/repokey.gpg'
```
Update the packages:
```
apt-get update && apt-get upgrade
```
Start the Services:
```
supervisorctl start all
```

Note(2): It is possible that a major version update will require extra changes, for doing that check the specific documentation for the version update.

> - Once the Primary Pod is update, it is necessary to do the same previous step for all the Secondary Pods one by one, doing this, each Secondary Pod will remain in "NotReady" state and will exit the Load Balancer redirect while it is being updated, but the rest will continue to work with the old version without disrupting the connection.

Reviewer 2: ```suggestion
- Once the Primary Pod is updated, it is necessary to do the same previous step for all the Secondary Pods one by one. While doing this, each Secondary Pod will remain in "NotReady" state and will exit the Load Balancer redirect until it is updated, but the rest will continue to work without disrupting the connection.
```

# 8 Message logs and disk space
As described in the [User Guide](https://github.com/nordic-institute/X-Road-Security-Server-sidecar/blob/master/doc/security_server_sidecar_user_guide.md#8-message-log) the recommend was to store the local storage of message log in a docker container. In Kubernetes we can create any of the Kubernetes volumes described in [4.2.3 Kubernetes Volumes](#423-Kubernetes-volumes).

> It is also recommend to send the logs inside the volume to a AWS S3 Bucket. For doing that it is required to:

Reviewer 2: ```suggestion
It is also recommended to send the logs inside the volume to an AWS S3 Bucket. To do that we need to:
```
